File: lib/javascript/src/time.ts
import { assertInRange, assertLength } from "./bytes.ts";

/** A time of day. */
export type Time = {
	/** `0` to `23`. */
	hour: number;
	/** `0` to `59`. */
	minute: number;
	/** `0` to `59`. */
	second: number;
};

/**
 * Decodes a time stored as three plain binary bytes.
 * @param bytes Hour, minute and second.
 */
export function decodeTime(bytes: Uint8Array): Time {
	assertLength(bytes, 3, "time");
	const time = { hour: bytes[0]!, minute: bytes[1]!, second: bytes[2]! };
	assertInRange("hour", time.hour, 0, 23);
	assertInRange("minute", time.minute, 0, 59);
	assertInRange("second", time.second, 0, 59);
	return time;
}

/** Encodes a time into three plain binary bytes. */
export function encodeTime({ hour, minute, second }: Time): Uint8Array {
	assertInRange("hour", hour, 0, 23);
	assertInRange("minute", minute, 0, 59);
	assertInRange("second", second, 0, 59);
	return Uint8Array.from([hour, minute, second]);
}

File: lib/rust/src/time.rs
use core::fmt;

use crate::bytes::in_range;
use crate::error::Result;

/// A time of day stored as three plain binary bytes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Time {
	/// `0` to `23`.
	pub hour: u8,
	/// `0` to `59`.
	pub minute: u8,
	/// `0` to `59`.
	pub second: u8,
}

impl Time {
	/// Decodes hour, minute and second bytes.
	///
	/// # Errors
	/// [`Error::Range`](crate::Error::Range) for an hour above 23 or a minute or second above 59.
	pub fn decode([hour, minute, second]: [u8; 3]) -> Result<Self> {
		Ok(Self {
			hour: in_range("hour", hour, 0, 23)?,
			minute: in_range("minute", minute, 0, 59)?,
			second: in_range("second", second, 0, 59)?,
		})
	}

	/// Encodes into hour, minute and second bytes.
	///
	/// # Errors
	/// [`Error::Range`](crate::Error::Range) for an hour above 23 or a minute or second above 59.
	pub fn encode(self) -> Result<[u8; 3]> {
		Ok([
			in_range("hour", self.hour, 0, 23)?,
			in_range("minute", self.minute, 0, 59)?,
			in_range("second", self.second, 0, 59)?,
		])
	}
}

impl fmt::Display for Time {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{:02}:{:02}:{:02}", self.hour, self.minute, self.second)
	}
}
